File: 程序员面试金典/3.5栈排序/SortedStack.java
class SortedStack {
    private Stack<Integer> mainStack, subStack;

    public SortedStack() {
        mainStack = new Stack<>();
        subStack = new Stack<>();
    }
    
    public void push(int val) {
        while (!mainStack.empty() && mainStack.peek() < val)
            subStack.push(mainStack.pop());
        while (!subStack.empty() && subStack.peek() > val)
            mainStack.push(subStack.pop());
        mainStack.push(val);
    }
    
    public void pop() {
        subToMain();
        if (!mainStack.empty())
            mainStack.pop();
    }
    
    public int peek() {
        subToMain();
        return mainStack.empty() ? -1 : mainStack.peek();
    }
    
    public boolean isEmpty() {
        return mainStack.empty() && subStack.empty();
    }

    private void subToMain() {
        while (!subStack.empty())
            mainStack.push(subStack.pop());
    }

}

/**
 * Your SortedStack object will be instantiated and called as such:
 * SortedStack obj = new SortedStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.isEmpty();
 */

File: 程序员面试金典/3.5/SortedStack.py
class SortedStack:

    def __init__(self):
        self.main_stack = []
        self.sub_stack = []

    def push(self, val: int) -> None:
        while self.main_stack and self.main_stack[-1] < val:
            self.sub_stack.append(self.main_stack.pop())
        self.main_stack.append(val)
        while self.sub_stack:
            self.main_stack.append(self.sub_stack.pop())

    def pop(self) -> None:
        if self.main_stack:
            self.main_stack.pop()

    def peek(self) -> int:
        return self.main_stack[-1] if self.main_stack else -1

    def isEmpty(self) -> bool:
        return len(self.main_stack) == 0


# Your SortedStack object will be instantiated and called as such:
# obj = SortedStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.peek()
# param_4 = obj.isEmpty()
